packages/imperative/src/auth/handlers/BaseAuthHandler.ts

```diff
@@ -0,0 +1,233 @@
+     */
+    protected abstract async doLogout(session: AbstractSession): Promise<void>;
+
+    /**
+     * Performs the login operation. Builds a session to connect to the auth
+     * service, sends a login request to it to obtain a token, and stores the
+     * resulting token in the profile of type `mProfileType`.
+     * @param {IHandlerParameters} params Command parameters sent by imperative.
+     */
+    private async processLogin(params: IHandlerParameters) {
+        const loadedProfile = params.profiles.getMeta(this.mProfileType, false);
+
+        const sessCfg: ISession = this.createSessCfgFromArgs(
+            params.arguments
+        );
+        const sessCfgWithCreds = await ConnectionPropsForSessCfg.addPropsOrPrompt<ISession>(
+            sessCfg, params.arguments,
+            { requestToken: true, defaultTokenType: this.mDefaultTokenType }
+        );
+
+        this.mSession = new Session(sessCfgWithCreds);
+
+        // login to obtain a token.
+        const tokenValue = await this.doLogin(this.mSession);
+
+        // update the profile given
+        if (loadedProfile != null && loadedProfile.name != null && !params.arguments.showToken && tokenValue != null) {
+            await Imperative.api.profileManager(this.mProfileType).update({
+                name: loadedProfile.name,
+                args: {
+                    "token-type": this.mSession.ISession.tokenType,
+                    "token-value": tokenValue
+                },
+                merge: true
+            });
+        } else if ((loadedProfile == null || loadedProfile.name == null ) && !params.arguments.showToken && tokenValue != null) {
+
+            // Do not store non-profile arguments, user, or password. Set param arguments for prompted values from session.
+
+            const copyArgs = {...params.arguments};
+            copyArgs.createProfile = undefined;
+            copyArgs.showToken = undefined;
+            copyArgs.user = undefined;
+            copyArgs.password = undefined;
+
+            copyArgs.host = this.mSession.ISession.hostname;
+            copyArgs.port = this.mSession.ISession.port;
+
+            copyArgs.tokenType = this.mSession.ISession.tokenType;
+            copyArgs["token-type"] = this.mSession.ISession.tokenType;
+
+            copyArgs.tokenValue = tokenValue;
+            copyArgs["token-value"] = tokenValue;
+
+            const createParms: ISaveProfileFromCliArgs = {
+                name: "default",
+                type: this.mProfileType,
+                args: copyArgs,
+                overwrite: false,
+                profile: {}
+            };
+
+            const answer: string = await CliUtils.promptWithTimeout(
+                "We require the host, port, and token must be stored on disk for use with future commands. Is this acceptable? [y/N]: ");
+            if (answer == null || !answer.toLowerCase().startsWith("y")) {
```

Comment:
is it safe to use `startsWith` or should we do `===`?

Reply:
My concern with === was that it would fail if a user entered "yes". I suppose we could just add logic to look for either "y" or "yes". I don't know why that didn't occur to me while I was writing this logic.

Reply:
Fixed.